src/deep-security/lib/Token.js

```diff
@@ -13,67 +15,113 @@ import {AuthException} from './Exception/AuthException';
     // @todo: set retries in a smarter way...
     AWS.config.maxRetries = 3;
 
-    let defaultRegion = AWS.config.region;
+    let cognitoRegion = Token.getRegionFromIdentityPoolId(this._identityPoolId);
 
     AWS.config.update({
-      region: Token._getRegionFromIdentityPoolId(this._identityPoolId),
+      cognitoidentity: {region: cognitoRegion},
+      cognitosync: {region: cognitoRegion},
     });
+  }
+
+  /**
+   * @returns {IdentityProvider}
+   */
+  get identityProvider() {
+    return this._identityProvider;
+  }
+
+  /**
+   * @param {IdentityProvider} provider
+   */
+  set identityProvider(provider) {
+    this._identityProvider = provider;
+  }
 
-    let cognitoParams = {
-      IdentityPoolId: this._identityPoolId,
-    };
+  /**
+   * @returns {Object}
+   */
+  get lambdaContext() {
+    return this._lambdaContext;
+  }
 
-    if (this._providerName && this._providerUserToken) {
-      this._isAnonymous = false;
-      cognitoParams.Logins = {};
-      cognitoParams.Logins[this._providerName] = this._providerUserToken;
+  /**
+   * @param {Object} lambdaContext
+   */
+  set lambdaContext(lambdaContext) {
+    this._lambdaContext = lambdaContext;
+  }
+
+  /**
+   * @param {Function} callback
+   */
+  loadCredentials(callback = () => null) {
+    // avoid refreshing or loading credentials for each request
+    if (this._validCredentials()) {
+      return callback(null, this.credentials);
     }
 
-    this._credentials = new AWS.CognitoIdentityCredentials(cognitoParams);
+    if (this.lambdaContext) {
+      this._credsManager.loadCredentials(this.identityId, (error, credentials) => {
+        if (error) {
+          return callback(error);
+        }
+
+        this._credentials = credentials;
 
-    // update AWS credentials
-    AWS.config.credentials = this._credentials.refresh(function(error) {
-      if (error) {
-        callback(new AuthException(error));
-        return;
+        callback(null, this._credentials);
+      });
+    } else {
+      let cognitoParams = {
+        IdentityPoolId: this._identityPoolId,
+      };
+
+      if (this.identityProvider) {
+        cognitoParams.Logins = {};
+        cognitoParams.Logins[this.identityProvider.name] = this.identityProvider.userToken;
       }
 
-      this._identityId = this._credentials.identityId;
+      this._credentials = new AWS.CognitoIdentityCredentials(cognitoParams);
 
-      AWS.config.update({
-        accessKeyId: this._credentials.accessKeyId,
-        secretAccessKey: this._credentials.secretAccessKey,
-        sessionToken: this._credentials.sessionToken,
-        region: defaultRegion, // restore to default region
-      });
+      this._credentials.refresh((error) => {
+        if (error) {
+          return callback(new AuthException(error));
+        }
+
+        AWS.config.credentials = this._credentials;
 
-      callback(null, this);
-    }.bind(this));
+        this._credsManager.saveCredentials(this._credentials, (error, record) => {
```

Comment:
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: ['record' is defined but never used.](https://www.codacy.com/app/MitocGroup/deep-framework/file/1363148763/issues/source?bid=2895475&fileBranchId=2973909#l116)